File: close_stale_prs.py
#!/usr/bin/env python3
"""
Script to close stale PRs and merge valid ones
"""

import json
import subprocess
import sys
import time

def run_command(cmd, capture_output=True):
    """Run a shell command and return the result"""
    try:
        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=capture_output,
            text=True,
            check=False
        )
        return result
    except Exception as e:
        print(f"Error running command: {e}")
        return None

def get_open_prs():
    """Get open PRs using GitHub CLI"""
    result = run_command("gh pr list --state open --json number,title,headRefName,mergeable,mergeStateStatus")
    if result and result.returncode == 0:
        try:
            return json.loads(result.stdout)
        except json.JSONDecodeError:
            return []
    return []

def check_branch_exists(branch_name):
    """Check if a branch exists in the remote repository"""
    result = run_command(f"git ls-remote --heads origin {branch_name}")
    return result and result.returncode == 0 and branch_name in result.stdout

def close_pr(pr_number, reason="Branch no longer exists"):
    """Close a PR with a reason"""
    print(f"🔒 Closing PR #{pr_number} - {reason}")
    
    # Close the PR
    close_result = run_command(f"gh pr close {pr_number} --comment 'Closing due to: {reason}'")
    if close_result and close_result.returncode == 0:
        print(f"✅ Successfully closed PR #{pr_number}")
        return True
    else:
        print(f"❌ Failed to close PR #{pr_number}: {close_result.stderr if close_result else 'Unknown error'}")
        return False

def merge_pr(pr_number):
    """Attempt to merge a PR"""
    print(f"🔄 Attempting to merge PR #{pr_number}...")
    
    # First, try to checkout the PR
    checkout_result = run_command(f"gh pr checkout {pr_number}")
    if checkout_result and checkout_result.returncode != 0:
        print(f"❌ Failed to checkout PR #{pr_number}: {checkout_result.stderr}")
        return False
    
    # Update the branch with latest main
    run_command("git fetch origin main")
    merge_result = run_command("git merge origin/main")
    if merge_result and merge_result.returncode != 0:
        print(f"⚠️  Merge conflicts in PR #{pr_number}, resolving...")
        # Try to resolve conflicts automatically
        run_command("git add .")
        run_command("git commit -m 'Resolve merge conflicts'")
    
    # Push any updates
    run_command("git push origin HEAD")
    
    # Try to merge
    merge_result = run_command(f"gh pr merge {pr_number} --merge --auto")
    if merge_result and merge_result.returncode == 0:
        print(f"✅ Successfully merged PR #{pr_number}")
        return True
    else:
        print(f"❌ Failed to merge PR #{pr_number}: {merge_result.stderr if merge_result else 'Unknown error'}")
        return False

def main():
    print("🚀 Starting PR cleanup and merge process...")
    
    # Get all open PRs
    prs = get_open_prs()
    if not prs:
        print("❌ No open PRs found")
        return 1
    
    print(f"Found {len(prs)} open PR(s)")
    
    # Sort by PR number (oldest first)
    prs.sort(key=lambda x: x['number'])
    
    success_count = 0
    closed_count = 0
    failed_count = 0
    
    for pr in prs:
        pr_number = pr['number']
        pr_title = pr['title']
        pr_branch = pr['headRefName']
        
        print(f"\n📋 Processing PR #{pr_number}: {pr_title}")
        print(f"   Branch: {pr_branch}")
        
        # Check if branch exists
        if not check_branch_exists(pr_branch):
            print(f"   ⚠️  Branch {pr_branch} no longer exists")
            if close_pr(pr_number, "Branch no longer exists"):
                closed_count += 1
            else:
                failed_count += 1
        else:
            # Try to merge the PR
            if merge_pr(pr_number):
                success_count += 1
            else:
                failed_count += 1
        
        # Small delay between operations
        time.sleep(1)
    
    print(f"\n🎉 Process completed!")
    print(f"   ✅ Successfully merged: {success_count}")
    print(f"   🔒 Closed stale PRs: {closed_count}")
    print(f"   ❌ Failed operations: {failed_count}")
    
    return 0 if failed_count == 0 else 1

if __name__ == "__main__":
    sys.exit(main())

File: close_stale_prs.js
#!/usr/bin/env node
import https from 'https';

// GitHub API configuration
const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
const REPO_OWNER = 'Zion-Holdings';
const REPO_NAME = 'zion.app';

// Function to make GitHub API requests
const makeGitHubRequest = (path, method = 'GET', data = null) => {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'api.github.com',
      path: path,
      method: method,
      headers: {
        'User-Agent': 'Zion-App-Automation',
        'Accept': 'application/vnd.github.v3+json',
        'Authorization': GITHUB_TOKEN ? `token ${GITHUB_TOKEN}` : ''
      }
    };

    if (data) {
      options.headers['Content-Type'] = 'application/json';
    }

    const req = https.request(options, res => {
      let responseData = '';
      res.on('data', chunk => {
        responseData += chunk;
      });
      res.on('end', () => {
        try {
          const jsonData = JSON.parse(responseData);
          resolve(jsonData);
        } catch (error) {
          reject(new Error(`Failed to parse JSON: ${responseData}`));
        }
      });
    });

    req.on('error', error => reject(error));

    if (data) {
      req.write(JSON.stringify(data));
    }
    req.end();
  });
};

// Function to check if a branch exists
async function checkBranchExists(branchName) {
  try {
    await makeGitHubRequest(`/repos/${REPO_OWNER}/${REPO_NAME}/branches/${branchName}`);
    return true;
  } catch (error) {
    return false;
  }
}

// Function to close a PR
async function closePR(pr) {
  try {
    await makeGitHubRequest(`/repos/${REPO_OWNER}/${REPO_NAME}/pulls/${pr.number}`, 'PATCH', {
      state: 'closed'
    });
    console.log(`  ✅ Closed PR #${pr.number} (${pr.title})`);
    return true;
  } catch (error) {
    console.log(`  ❌ Failed to close PR #${pr.number}: ${error.message}`);
    return false;
  }
}

// Main function
async function main() {
  try {
    console.log('🔍 Checking for stale PRs...');
    
    // Get all open PRs
    const prs = await makeGitHubRequest(`/repos/${REPO_OWNER}/${REPO_NAME}/pulls?state=open&per_page=100`);
    
    if (prs.length === 0) {
      console.log('✅ No open PRs found');
      return;
    }
    
    console.log(`📊 Found ${prs.length} open PR(s)`);
    
    // Filter PRs that target main branch
    const mainPRs = prs.filter(pr => pr.base.ref === 'main');
    console.log(`🎯 ${mainPRs.length} PR(s) targeting main branch`);
    
    let closedCount = 0;
    let skippedCount = 0;
    
    // Process each PR
    for (const pr of mainPRs) {
      console.log(`\n🔍 Checking PR #${pr.number}: ${pr.title}`);
      console.log(`  Branch: ${pr.head.ref}`);
      
      // Check if the branch still exists
      const branchExists = await checkBranchExists(pr.head.ref);
      
      if (!branchExists) {
        console.log(`  ⚠️  Branch ${pr.head.ref} no longer exists, closing PR...`);
        const closed = await closePR(pr);
        if (closed) {
          closedCount++;
        }
      } else {
        console.log(`  ✅ Branch ${pr.head.ref} still exists, keeping PR open`);
        skippedCount++;
      }
      
      // Small delay between requests
      await new Promise(resolve => setTimeout(resolve, 500));
    }
    
    console.log(`\n📊 Close Summary:`);
    console.log(`  🔒 Closed stale PRs: ${closedCount}`);
    console.log(`  ⏭️  Skipped active PRs: ${skippedCount}`);
    console.log(`  📋 Total processed: ${mainPRs.length}`);
    
  } catch (error) {
    console.error('❌ Error in main process:', error.message);
  }
}

main().catch(console.error);